srml/operator/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+		pub OperatorHasContracts: map T::AccountId => BTreeSet<T::AccountId>;
+		/// A mapping from operated contract by operator to it.
+		pub ContractHasOperator: map T::AccountId => Option<T::AccountId>;
+		/// A mapping from contract to it's parameters.
+		pub ContractParameters: map T::AccountId => Option<T::Parameters>;
+	}
+}
+
+decl_module! {
+	/// The module declaration.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		// Initializing events
+		// this is needed only if you are using events in your module
+		fn deposit_event() = default;
+
+		/// Deploys a contact and insert relation of a contract and an operator to mapping.
+		pub fn instantiate(origin,
+			#[compact] endowment: BalanceOf<T>,
+			#[compact] gas_limit: Gas,
+			code_hash: CodeHash<T>,
+			data: Vec<u8>,
+			parameters: T::Parameters) -> Result {
+			let operator = ensure_signed(origin)?;
+
+			// verify parameters.
+			parameters.verify()?;
+
+			let contract = T::DetermineContractAddress::contract_address_for(&code_hash, &data, &operator);
+			contract::Module::<T>::instantiate(RawOrigin::Signed(operator.clone()).into(), endowment, gas_limit, code_hash, data)?;
+
+			// add operator to contracts
+			<OperatorHasContracts<T>>::mutate(&operator, {|tree| (*tree).insert(contract.clone()) });
+			// add contract to operator
+			<ContractHasOperator<T>>::insert(&contract, operator.clone());
+			// add contract to parameters
+			<ContractParameters<T>>::insert(&contract, parameters);
+
+			// issue an event operator -> contract
+			Self::deposit_event(RawEvent::SetOperator(operator, contract));
+			Ok(())
+		}
+
+		/// Updates parameters for an identified contact.
+		pub fn update_parameters(origin, contract: T::AccountId, parameters: T::Parameters) -> Result {
+			let operator = ensure_signed(origin)?;
+
+			// verify parameters
+			parameters.verify()?;
+
+			let contracts = <OperatorHasContracts<T>>::get(&operator);
+
+			// check the actually operate the contract.
+			if !contracts.contains(&contract) {
+				return Err("The sender don't operate the contract address.")
+			}
+
+			// update parameters
+			<ContractParameters<T>>::insert(&contract, parameters.clone());
+			// issue set parameter events
+			Self::deposit_event(RawEvent::SetParameters(contract, parameters));
+			Ok(())
+		}
+
+		/// Changes an operator for identified contracts.
+		pub fn change_operator(origin, contracts: Vec<T::AccountId>, new_operator: T::AccountId) -> Result {
+			let operator = ensure_signed(origin)?;
+			let operate_contracts = <OperatorHasContracts<T>>::get(&operator);
+
+			// check the actually operate the contract.
+			for c in contracts.iter() {
+				if !operate_contracts.contains(&c) {
+					return Err("The sender don't operate the contracts address.")
+				}
+			}
+
+
+			for c in contracts.iter() {
+				// remove origin operator to contracts
+				<OperatorHasContracts<T>>::mutate(&operator, {|tree| (*tree).remove(c) });
```

Comment:
This `mutate` operation looks a bit expensive, whats about optimize mutate calls like

```rust
<OperatorHasContracts<T>>::mutate(&operator,
    |tree| for c in contracts.iter() { (*tree).remove(c); }
);
```

Reply:
Thanks, so goods!